src/Encoding.chpl

```diff
@@ -0,0 +1,254 @@
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:uint(64);
+  }
+
+  proc uint32Append(val: uint(32), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc uint32Consume(ch: channel(false,iokind.little,false)): uint(32) throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:uint(32);
+  }
+  
+
+  proc int64Append(val: int(64), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc int64Consume(ch: channel(false,iokind.little,false)): int throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:int(64);
+  }
+
+  proc int32Append(val: int(32), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc int32Consume(ch: channel(false,iokind.little,false)): int(32) throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:int(32);
+  }
+
+  proc boolAppend(val: bool, ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc boolConsume(ch: channel(false,iokind.little,false)): bool throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:bool;
+  }
+
+  proc sint64Append(val: int(64), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = (val << 1):uint ^ (val >> 63):uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc sint64Consume(ch: channel(false,iokind.little,false)): int(64) throws {
+    const (val, len) = unsignedVarintConsume(ch);
+    return (val >> 1):int(64) ^ (val):int(64) << 63 >> 63;
+  }
+
+  proc sint32Append(val: int(64), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = (val << 1):uint ^ (val >> 31):uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc sint32Consume(ch: channel(false,iokind.little,false)): int(32) throws {
+    const (val, len) = unsignedVarintConsume(ch);
+    return (val >> 1):int(32) ^ (val):int(32) << 31 >> 31;
+  }
+
+  proc bytesAppend(val: bytes, ch: channel(true,iokind.little,false)) throws {
+    const wireType = lengthDelimited;
+    unsignedVarintAppend((val.size):uint, ch);
+    ch.write(val);
+  }
+
+  proc bytesConsume(ch: channel(false,iokind.little,false)): bytes throws {
+    const (byteLen, len) = unsignedVarintConsume(ch);
+    var s:bytes;
+    ch.readbytes(s, byteLen:int);
+    return s;
+  }
+
+  proc stringAppend(val: string, ch: channel(true,iokind.little,false)) throws {
+    bytesAppend(val.encode(), ch);
+  }
+
+  proc stringConsume(ch: channel(false,iokind.little,false)): string throws {
+    return bytesConsume(ch).decode();
```

Comment:
This will throw if the string is not UTF-8. Is that expected behavior for protocol buffer messages? (If this is a complex question to answer, feel free to add a TODO to check in your task list).

Reply:
> This will throw if the string is not UTF-8

Can you please give an example for which it throws, I tried with something like "~𝘈Ḇ𝖢𝕯٤ḞԍНǏ𝙅ƘԸⲘ𝙉০Ρ𝗤Ɍ𝓢ȚЦ", and it passes both general tests and end to end tests.

Reply:
From the protobuf encoding guide for [Strings](https://developers.google.com/protocol-buffers/docs/encoding#strings)

> The red bytes are the UTF8 of "testing". The key here is 0x12 →

I understand that string type expects users to pass a UTF-8 value, other values can be handles through bytes. Though I am not 100% sure on this.

Reply:
I think protocol buffer strings are always UTF-8 - see https://developers.google.com/protocol-buffers/docs/proto3#scalar 

>   A string must always contain UTF-8 encoded or 7-bit ASCII text, and cannot be longer than 2^32.

Reply:
Yes! Agreed.